src/tasks/TaskHelper.ts

```diff
@@ -0,0 +1,166 @@
+    resolveDockerRunOptions(context: DockerRunTaskContext, runDefinition: DockerRunTaskDefinition): Promise<DockerRunOptions>;
+}
+
+export function registerTaskProviders(ctx: ExtensionContext): void {
+    const helpers = {
+        netCore: netCoreTaskHelper,
+        node: nodeTaskHelper
+    };
+
+    ctx.subscriptions.push(
+        tasks.registerTaskProvider(
+            'docker-build',
+            new DockerBuildTaskProvider(helpers)
+        )
+    );
+
+    ctx.subscriptions.push(
+        tasks.registerTaskProvider(
+            'docker-run',
+            new DockerRunTaskProvider(helpers)
+        )
+    );
+}
+
+export async function addTask(task: DockerBuildTaskDefinition | DockerRunTaskDefinition): Promise<boolean> {
+    // Using config API instead of tasks API means no wasted perf on re-resolving the tasks, and avoids confusion on resolved type !== true type
+    const workspaceTasks = workspace.getConfiguration('tasks');
+    const allTasks = workspaceTasks && workspaceTasks.tasks as TaskDefinition[] || [];
+
+    if (allTasks.some(t => t.label === task.label)) {
+        return false;
+    }
+
+    allTasks.push(task);
+    await workspaceTasks.update('tasks', allTasks);
+    return true;
+}
+
+export async function getAssociatedDockerRunTask(debugConfiguration: DockerDebugConfiguration): Promise<DockerRunTaskDefinition | undefined> {
+    // Using config API instead of tasks API means no wasted perf on re-resolving the tasks, and avoids confusion on resolved type !== true type
+    const workspaceTasks = workspace.getConfiguration('tasks');
+    const allTasks: TaskDefinition[] = workspaceTasks && workspaceTasks.tasks as TaskDefinition[] || [];
+
+    return await recursiveFindTaskByType(allTasks, 'docker-run', debugConfiguration);
+}
+
+export async function getAssociatedDockerBuildTask(runTask: DockerRunTaskDefinition): Promise<DockerBuildTaskDefinition | undefined> {
+    // Using config API instead of tasks API means no wasted perf on re-resolving the tasks, and avoids confusion on resolved type !== true type
+    const workspaceTasks = workspace.getConfiguration('tasks');
+    const allTasks: TaskDefinition[] = workspaceTasks && workspaceTasks.tasks as TaskDefinition[] || [];
+
+    return await recursiveFindTaskByType(allTasks, 'docker-build', runTask);
+}
+
+export async function getOfficialBuildTaskForDockerfile(dockerfile: string, folder: WorkspaceFolder): Promise<Task | undefined> {
+    let buildTasks = await tasks.fetchTasks({ type: 'docker-build' });
+    buildTasks =
+        buildTasks.filter(t => t.execution.args.some(a => { // Find all build tasks where an argument to 'docker build' is this Dockerfile
+            let arg: string;
+            if (typeof a === 'string') {
+                arg = a;
+            } else {
+                arg = a.value;
+            }
+
+            arg = resolveWorkspaceFolderPath(folder, arg);
+            return arg === dockerfile;

```

Comment:
>arg [](start = 19, length = 3)

Intentionally case-sensitive file path comparison?

Reply:
Yes. Most of our users are on Unix. If they have incorrect casing in what they specify, it won't work for them anyway.

Reply:
macOS and Windows use (by default) case-insensitive file path comparison.  Those platforms are 80% of our usage.

Reply:
Very well. Made it case-insensitive.